### PR TITLE
feat(replays): add download functionality to replay recording segment index

### PR DESCRIPTION
**Pull request by JoshFerge**
Right now the frontend has to make an individual request to download every single recording segment on the details page. It's possible/common that there's over 100 segments for a recording, which means the frontend has to make _a lot_ of requests.

This PR adds functionality to the index endpoint so that it can be used to download a range of segments, using our standard pagination parameters / headers. The endpoint will also deflate the data and return it as JSON, which will remove the need for us to include `pako` client side. 

We can experiment with optimal loading of a replay details page by fetching only the first few segments and then the rest, or simply batching requests using a fixed offset (say grab 20 recording segments per request). 

We could also try including metadata about `checkout` segments, and fetch from checkout to checkout, but this will be more involved since that requires a client sdk change.

### CODE REVIEW

File: tests/sentry/replays/test_project_replay_recording_segment_index.py [hunk 1]
```diff
@@ -54,3 +56,95 @@ def test_index(self):
...
+            f.putfile(BytesIO(f'{{"test":"hello {i}"}}'.encode()))
+            ReplayRecordingSegment.objects.create(
+                replay_id=self.replay_id,
+                project_id=self.project.id,
+                segment_id=i,
+                file_id=f.id,
+            )
+
+        with self.feature("organizations:session-replay"):
+            response = self.client.get(self.url + "?download=true")
```
Comment (cmanallen) [inline tests/sentry/replays/test_project_replay_recording_segment_index.py:108]: Very nitty.  Little disconnected between endpoint and coverage here.  The endpoint expects `?download` is specified.  It doesn't care about its value.

File: src/sentry/replays/endpoints/project_replay_recording_segment_index.py [hunk 2]
```diff
@@ -18,15 +26,80 @@ def get(self, request: Request, project, replay_id: str) -> Response:
...
+        for i, file in enumerate(recording_segments):
+            if self.is_compressed(file):
+                yield from self.decompress_blob_stream(file)
+            else:
+                yield file.read().decode("utf-8")
+
+            if i < len(recording_segments) - 1:
+                yield ","
+
+        yield "]"
```
Comment (cmanallen) [inline src/sentry/replays/endpoints/project_replay_recording_segment_index.py:89]: Do we need to serialize this as JSON?

**Review by cmanallen — APPROVED** [review]

File: src/sentry/replays/endpoints/project_replay_recording_segment_index.py (hunk 2 above)
Comment (JoshFerge) [reply src/sentry/replays/endpoints/project_replay_recording_segment_index.py:89]: if we don't do it here the client has to its own parsing / serialization, and so it here can clean up that client side logic.

File: tests/sentry/replays/test_project_replay_recording_segment_index.py [hunk 3]
```diff
@@ -54,3 +56,95 @@ def test_index(self):
...
+
+        assert response.status_code == 200
+
+        assert response.get("Content-Type") == "application/json"
+        assert b'[{"test":"hello 0"},{"test":"hello 1"},{"test":"hello 2"}]' == b"".join(
```
Comment (cmanallen) [inline tests/sentry/replays/test_project_replay_recording_segment_index.py:92]: Can you add a test case where the value is compressed and exceeds 4096 bytes?

Comment (JoshFerge) [reply tests/sentry/replays/test_project_replay_recording_segment_index.py:92]: added

**Pull request merged**